### PR TITLE
ppu: Lwmutex, lwcond and scheduler fixes

### PR DESCRIPTION
**Pull request by elad335**
- Fix lv2 lwmutex locking, use a simple flag to tell its state, thats how the firmware does it. Fixes a few assertation fails when there is an unusual use of the lock such as unlocking twice etc. 

- Fix thread scheduler: Properly check if running thread is already queued before inserting it. Fixes a few "deadly" thread notifications.

- Dont reaqcuire lock if timeout passed in lv2 lwcond queue_wait; [testcase](https://github.com/elad335/myps3tests/tree/master/ppu_tests/lv2%20lwcond%20wait%20timeout)

Brings [Le Tour de France 2012](https://forums.rpcs3.net/thread-199034.html) from loadable into intro.

image

### CODE REVIEW

**Comment by stride21** [comment]
Motorstorm no longer has race conditions in the main menu and is now able to get ingame when selecting the first race.
motorstorm

**Comment by jenci8888** [comment]
Gran Turismo 5 Prologue possible go ingame (everything expect PSN JPN but needs more test)
image

Gran Turismo 5 (only intro at moment because broken graphics and infinite loading when i click something random like settings, track editor, home, GT championship etc.)
but first you need to delete save data before launch game otherwise it will stuck loadable
image

Gran Turismo Academy 2013 with rendered suzuka track in login screen (compiled 722 shaders) but it requires login PSN account, so we can't do anything
image

(More further testing, tomorrow i'll test Gran Turismo 6)

**Comment by micaelis72** [comment]
Gundam Breaker 2 [BLAS50770] has some errors with lwcond

Logs, as usual

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2493961/RPCS3.log.gz)

**Comment by legend800** [comment]
Fixes: 
- Confirm Motorstorm can get ingame now. But soon freezes after entering gameplay. Playable speed though. Fix the freeze and minor gfx issues and this game is fully playable!

- Backbreaker and Backbreaker Vengeance are now perfectly stable! Thank you! Fixes this probably:  https://github.com/RPCS3/rpcs3/issues/3634

Regression:
- Splatterhouse has always been unstable on PPU LLVM and crashes at the beginning of stages (after lvl load). However with this PR, it crashes sooner (during the load screens). The good news may be though that you may be in the area where the original crashes are being triggered?

F {PPU[0x1000010] Thread (3023bec0) [0x01655ae0]} class std::runtime_error thrown: Verification failed: 
(in file c:\projects\rpcs3\rpcs3\emu\cell\lv2\sys_lwcond.cpp:293)

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2494505/RPCS3.log.gz)

**Comment by MarioSonic2987** [comment]
Skate Demo now goes Ingame without restarting the emulator many times due to this:
```
F {PPU[0x1000007] Thread (rwfilesys_sk8) [0x00a4af30]} class std::runtime_error thrown: Verification failed (e=0x7e): 
(in file c:\projects\rpcs3\rpcs3\emu\cell\lv2\sys_cond.cpp:279)
```

**Before** (crashes in the title screen or after):
image

**After** (it goes ingame all times):
image

Tested the game 3 times.

**Comment by MSuih** [comment]
Fixes the full version of Skate 1 as well. In master it crashes with  `'sys_mutex_unlock' failed with 0x80010009 : CELL_EPERM [1]` or `Verification failed (e=0x3): 
(...\Emu\Cell\lv2\sys_cond.cpp:279)` when loading main menu but this one can get past that and in game. I completed some missions as well and framerate seemed fairly solid

File: rpcs3/Emu/Cell/lv2/sys_lwmutex.h [hunk 1]
```diff
@@ -58,7 +58,7 @@ struct lv2_lwmutex final : lv2_obj
 	const u64 name;
 
 	semaphore<> mutex;
-	atomic_t<u32> signaled{0};
+	atomic_t<bool> locked{true};
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/sys_lwmutex.h]: Who called it "locked"?

File: Utilities/Thread.cpp [hunk 2]
```diff
@@ -1865,9 +1865,8 @@ void thread_base::detach()
 
 void thread_base::notify()
 {
-	if (!(m_signal & 1))
```
Comment (Nekotekina) [inline Utilities/Thread.cpp]: Not a bug, not a fix.

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 3]
```diff
@@ -1063,7 +1063,7 @@ void lv2_obj::awake(cpu_thread& cpu, u32 prio)
 
 	std::lock_guard lock(g_mutex);
 
-	if (prio == -4)
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Pointless changes.

File: Utilities/Thread.cpp (hunk 2 above)
Comment (elad335) [reply Utilities/Thread.cpp]: if flag turned on between the if and the OR on other thread, it didnt detect it.

File: rpcs3/Emu/Cell/lv2/sys_lwmutex.h (hunk 1 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_lwmutex.h]: ?

File: Utilities/Thread.cpp (hunk 2 above)
Comment (Nekotekina) [reply Utilities/Thread.cpp]: Not a bug, not a fix. Don't make me repeat myself, or I'll just close the PR.

**Comment by elad335** [comment]
I've removed the long lived cellsurmixer thread hack, as testing implies the its race condition is fixed in this pr. for testers: test the more games you can and check for issues (crashes, audio loss etc). If the smallest issue appears I'll drop the commit.

**Comment by legend800** [comment]
Regression fixed in Splatterhouse - loads gameplay now before freezing. :)
F {PPU[0x1000000] Thread (main_thread) [0x003c6488]} MEM: Access violation reading location 0x11e

File: rpcs3/Emu/Cell/lv2/sys_lwcond.cpp [hunk 4]
```diff
@@ -73,11 +73,11 @@ error_code _sys_lwcond_signal(ppu_thread& ppu, u32 lwcond_id, u32 lwmutex_id, u3
 		fmt::throw_exception("Unknown mode (%d)" HERE, mode);
 	}
 
-	lv2_lwmutex* mutex = nullptr;
+	lv2_lwmutex* mutex;
 
 	const auto cond = idm::check<lv2_obj, lv2_lwcond>(lwcond_id, [&](lv2_lwcond& cond) -> cpu_thread*
 	{
-		mutex = idm::check_unlocked<lv2_obj, lv2_lwmutex>(lwmutex_id);
+		if (!(mutex = idm::check_unlocked<lv2_obj, lv2_lwmutex>(lwmutex_id))) return 0;
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/sys_lwcond.cpp]: Check it on the next line please.

File: rpcs3/Emu/Cell/lv2/sys_lwmutex.cpp [hunk 5]
```diff
@@ -75,22 +75,16 @@ error_code _sys_lwmutex_lock(ppu_thread& ppu, u32 lwmutex_id, u64 timeout)
 
 	const auto mutex = idm::get<lv2_obj, lv2_lwmutex>(lwmutex_id, [&](lv2_lwmutex& mutex)
 	{
-		if (u32 value = mutex.signaled)
+		if (mutex.signaled.exchange(0))
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/sys_lwmutex.cpp]: Use new try_dec method here, it should be better.

File: rpcs3/Emu/Cell/lv2/sys_lwcond.cpp [hunk 6]
```diff
@@ -290,12 +290,6 @@ error_code _sys_lwcond_queue_wait(ppu_thread& ppu, u32 lwcond_id, u32 lwmutex_id
 
 				cond->waiters--;
 
-				if (mutex->signaled.fetch_dec_sat())
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/sys_lwcond.cpp]: What problems was it causing?
The change itself is acceptable though.

Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_lwcond.cpp]: This commit originated from glances at the syscall in ida, so I've tested this using direct lv2 calls to test the mutex state after the timeout. eg the code below would generate correct results with this change.
```
inline bool test_lock(sys_lwmutex_t* mtx)
{
	if (_sys_lwmutex_trylock(mtx) == CELL_OK) 
	{
		_sys_lwmutex_unlock(mtx);
		return false;
	}
	return true;
}
```

Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_lwcond.cpp]: those are syscall macros from the testcase code.

**Comment by digitaldude555** [comment]
screenshot 544

Also improves Tour de France 2014 from loadable to intro.

**Comment by legend800** [comment]
@elad335 Why was this not merged, along with the other ~8 PRs of yours? 

Can I at least get a build of this PR as it made it made Backbreaker fully playable and I didn't save a build?

**Comment by elad335** [comment]
@legend800 its on appveyor.

**Comment by RainbowCookie32** [comment]
> Will this be merged to master eventually? I'd love to see Skate 1 playable and this provides that according to a few people who have tested it.

You can still download the build and use it

**Comment by HexyFify** [comment]
@ipkpjersi it will be merged when the code is correct

**Pull request closed without merging**